### PR TITLE
Extend internal API docs, part 2

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

The internal API docs now contain more useful information.

# Context

Picking up from #8111.

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

File: src/libstore/store-cast.hh [hunk 1]
```diff
@@ -4,6 +4,13 @@
 
 namespace nix {
 
+/**
+ * Helper to try downcasting a Store with a nice method if it fails.
+ *
+ * This is basically an alternative to the user-facing part of
+ * Store::unsupported that llows us to still have a nice message but
```
Comment (edolstra) [inline src/libstore/store-cast.hh:11]:
```suggestion
 * Store::unsupported that allows us to still have a nice message but
```

File: src/libutil/archive.hh [hunk 2]
```diff
@@ -7,54 +7,71 @@
...
+ *
+ *     IF path points to a DIRECTORY:
+ *       dump(path) = attrs(
+ *         [ ("type", "directory")
+ *         , ("entries", concat(map(f, sort(entries(path)))))
+ *         ])
+ *         where f(fn) = attrs(
+ *           [ ("name", fn)
+ *           , ("file", dump(path + "/" + fn))
+ *           ])
```
Comment (edolstra) [inline src/libutil/archive.hh:36]: How does this get formatted?

File: src/libutil/comparator.hh [hunk 3]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
-/* Awfull hacky generation of the comparison operators by doing a lexicographic
+/**
+ * Awfull hacky generation of the comparison operators by doing a lexicographic
```
Comment (edolstra) [inline src/libutil/comparator.hh:4]:
```suggestion
 * Awful hacky generation of the comparison operators by doing a lexicographic
```

File: src/libutil/archive.hh (hunk 2 above)
Comment (Ericson2314) [reply src/libutil/archive.hh:37]: Hehe so it actually wasn't rendering at all, but then I fixed that in the new second commit, and it is rendering correctly in a code block.

**Comment by Ericson2314** [comment]
I think I want to be a bit more aggressive merging this thing, as to quote @thufschmitt 

> Thanks, I think it's great to have this.
>
> I didn't check every comment in detail as I believe it's OK for these to be imperfect at first, just noted the things that stood out.

I.e. even if I did something silly inside PRs, it's probably still a clear net positive as no behavior is changed and no code is refactored. We have plenty of other PRs more deserving of thorough reviews!

**Pull request merged**